### PR TITLE
New `www` page to explain mocking

### PR DESCRIPTION
**Pull request by jharlowuk**
Hey folks! Thank you so much for creating tRPC, I'm sure you don't need more compliments but I am using it pretty much everyday and completely adore what it's done for my productivity 😄 

## 🎯 Changes
- This PR aims to add a new page called `Mocking` that goes through the steps needed to mock tRPC calls from components being rendered in isolated environments (tests, Storybook, etc.)
- The page goes through why mocking at the Service Worker level is worthwhile, how to create a `TRPCMockedProvider` component that can be used in various context, and how to write performant, easy to execute frontend unit tests on React components that contain tRPC queries and mutations

## 👀 Why I wrote this
In setting up my own tRPC client and server in Next.js, getting set up with unit tests has taken a large amount of time. I understand the appeal of having a local backend to test on, however, I felt like there are enough people asking for mocking and mocking is a standard and good enough practice that having a mocking page would be valuable to those looking to do it. 

I've also tried to make the page focused enough on "mocking" nomenclature that I don't think someone looking specifically to an integration or E2E test would find this page first. I hope this is acceptable! If not, I'm happy to find somewhere else to post this ❤️ 

## ✅ Checklist
- [X] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: www/docs/client/mocking.md
Comment (c-ehrlich) [inline www/docs/client/mocking.md:24]: This section reads a bit confusing to me. Here's my attempt:

In order for tRPC's hooks to work, they require context that describes the URL of the backend and the structure of the API.

- In React, you achieve this by wrapping your component tree with the `trpc.Provider` and `QueryClientProvider` components.
- In Next.js, this step is abstracted in `_app.ts` by using `trpc.withTRPC(MyApp)`.

We still need this context when using our components in a testing environment, but we want the requests to be made to a mocked version of our API as this allows us test the frontend in isolation. For this, we will need a mocked provider.

File: www/docs/client/mocking.md
Comment (c-ehrlich) [inline www/docs/client/mocking.md:28]:
```suggestion
Create a new file for your `MockedProvider` and import your `AppRouter` type from where your standard tRPC router is defined. This will ensure that your mocks have the same type signature as your API.
```

Also there seems to be some inconsistency between `MockedProvider` and `trpcMockedProvider`. I think these are referring to the same thing?

File: www/docs/client/mocking.md
Comment (c-ehrlich) [inline www/docs/client/mocking.md:34]:
```suggestion
Next, create a set of React hooks from your `AppRouter` type signature with `createTRPCReact`.
```
not sure the second half of the sentence adds anything here

File: www/docs/client/mocking.md
Comment (c-ehrlich) [inline www/docs/client/mocking.md:51]: Maybe link to the specific equivalent part of the React docs? (https://trpc.io/docs/react#3-add-trpc-providers)

File: www/docs/client/mocking.md
Comment (c-ehrlich) [inline www/docs/client/mocking.md:171]: I feel like showing this incorrect example first isn't helpful. Maybe just go straight to the correct one, and explain why it's correct?

File: www/docs/client/mocking.md
Comment (c-ehrlich) [inline www/docs/client/mocking.md:17]: I personally prefer `MockedTRPCProvider` as it makes it more obvious what it is, but don't feel strongly about this

File: www/docs/client/mocking.md
Comment (c-ehrlich) [inline www/docs/client/mocking.md:113]:
```suggestion
To utilize tRPC's typing superpowers, we can create a `mswTrpc` constant which generates a typed set of handlers for MSW, based on the shape of your `AppRouter`. This enables us to build handlers specific to our individual tests rapdily and with typesafety.
```

File: www/docs/client/mocking.md
Comment (c-ehrlich) [inline www/docs/client/mocking.md:126]: There is probably a more clear phrasing than "defining them to your network calls" but I can't figure out what it is right now either

File: www/docs/client/mocking.md
Comment (c-ehrlich) [inline www/docs/client/mocking.md:132]:
```suggestion
Let's say that we recently defined an endpoint in our `cats` router to serve everyone with information about our favorite cat.
```

File: www/docs/client/mocking.md
Comment (c-ehrlich) [inline www/docs/client/mocking.md:159]: I think the way you've written this component, data would still be nullable at this point?

Here's an example of how this is done in a different docs page: https://trpc.io/docs/react#4-fetch-data
tbh I don't like that version either as it shows "loading" on error, but maybe a reasonable compromise between correctness and not being too many lines of code to distract from what the actual topic is

**Review by c-ehrlich — COMMENTED** [review]
Looks good overall, and how to mock is a sufficiently common question that it feels worth having a docs page imo.

General notes:
- You currently haven't registered the page to actually show up in the docs. I think it could make sense as one page in a larger "testing" section (other pages could be testing procedures with a caller, and e2e tests)
- this page is currently strictly about React. I'm leaning towards this being ok, but wanted to point it out
- `msw-trpc` is still really new. Are we ok to have it in the docs as an "officially endorsed" solution? I'm leaning yes on this one as well.
- While I think it's good to have documentation about testing, I personally try to structure my applications in a way where I don't need to mock much at all, by using [Presentational and Container components](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0) to the point where the only thing a mock would really test is that `fetch` works. iirc @KATT isn't too fond of excessive unit tests with mocking either? I understand that it can sometimes be necessary, but maybe we can start by guiding people on how to avoid it when it's not necessary?

**Comment by Nick-Lucas** [comment]
> * this page is currently strictly about React

Yes this should not be under `client/`, it shouldn't be `reactjs/` and attached to that category in the sidebar config :)

**Comment by jharlowuk** [comment]
Thanks folks, I'm going to take a bit of time to rewrite and address your comments. I appreciate such in depth and detailed feedback @c-ehrlich and agree with everything you've mentioned.

File: www/docs/client/mocking.md
Comment (jharlowuk) [reply www/docs/client/mocking.md:24]: Thank you. I like your version a lot better.

I've made two minor, additional tweaks:
- I've added a link to our documentation on context, as I think that if you don't have context on what `context` is, this sentence could be confusing. 
- I've played around the the phrasing of the last paragraph to make it clear that our mocked provider is enabling the component to make requests as it normally would and that mocking the API is something we'll do later. It's minor change but I think this illustrates the difference that this current step is making a little more clearly

> In order for tRPC's hooks to work, they require [context](https://trpc.io/docs/context) that describes the URL of the backend and the structure of the API.

> - In React, you achieve this by wrapping your component tree with the `trpc.Provider` and `QueryClientProvider` components.
> - In Next.js, this step is abstracted in `_app.ts` by using `trpc.withTRPC(MyApp)`.

> We still need this context when using our components in an isolated environment, because they still need to be capable of making requests that we will later intercept and pass through the mocked version of our API. For this, we will need to create and use a mocked provider.

File: www/docs/client/mocking.md
Comment (jharlowuk) [reply www/docs/client/mocking.md:28]: Agreed! Thank you I have updated all the naming to be more consistent and follow the styling rule that:
- `Mocked` is always the first word if it is included
- `TRPC` is always capitalized or (lower case if at the start of a variable name)

File: www/docs/client/mocking.md
Comment (jharlowuk) [reply www/docs/client/mocking.md:34]: I wrote this because I was thinking about the fact folks using Next.js should still use `createTRPCReact` at this step, instead of `createTRPCNext`.

I'm going to delete the sentance because I agree with you it doesn't affect any of the steps someone should take and will consider adding an `:::info` to highlight this earlier in the page. I think that will also address some of the feedback re: placing this in `/react`

File: www/docs/client/mocking.md
Comment (jharlowuk) [reply www/docs/client/mocking.md:113]: Sweet, thank you!

File: www/docs/client/mocking.md
Comment (jharlowuk) [reply www/docs/client/mocking.md:126]: Agreed. I've tried making it more explicit, here's my rephrasing:

> If you want to skip this step, it is possible to create standard REST handlers using `msw`s `rest` object by manually defining the Request URL and Method corresponding to the tRPC procedure you want to mock.

File: www/docs/client/mocking.md
Comment (jharlowuk) [reply www/docs/client/mocking.md:159]: You are correct, I think there's always going to be that compromise, but I agree with the sentiment, especially as you get a linter error for an unused variable from `error` anyway. Here's my shot at handling this without adding extra lines:

```
export default const BestCat = () => {
  const { loading, error, data } = trpc.cats.bestCat.useQuery({ name: "John" })
  if (loading || error) return <p>Data unavaliable</p>;
  return (
    <StyledCatText>{data.bestCat}</StyledCatText>
  )
}
```

File: www/docs/client/mocking.md
Comment (jharlowuk) [reply www/docs/client/mocking.md:15]: Sound good, updated to Vitest in this line and I believe that the only difference is that `it` should now be `test`. Would appreciate a second eye on that though as I am unfamiliar with Vitest

**Comment by bautistaaa** [comment]
thanks for the thorough review @c-ehrlich 

great PR @jharlowuk

File: www/docs/client/mocking.md
Comment (jharlowuk) [reply www/docs/client/mocking.md:59]: You are completely correct! I've removed this line.

File: www/docs/client/mocking.md
Comment (jharlowuk) [reply www/docs/client/mocking.md:60]: What it means is that when you define request handlers with `msw-trpc`, it is only listening for the specific HTTP RPC endpoint for the intended endpoint you specify from the router. When you batch things, it doesn't create any errors in this instance, however, a request handler won't exist in `msw`'s "server" to intercept the resulting HTTP call.

In this instance, you'll see a bright yellow warning in your terminal (assuming you are executing a test) letting you know that the component sent a request that wasn't intercepted. Honestly, I've struggled to get a `mockedTRPCClient` running without specifying a batch link. Would appreciate your insight if I'm doing something goofy.

**Comment by jharlowuk** [comment]
Hey folks,

I've updated the doc to reflect your feedback and moved it into the `/reactjs` folder. Would appreciate another round of feedback based on my comments where I've left them unresolved 👍 

Re: @c-ehrlich's general notes:
- I'm happy to work on a testing page at some point, but in the interest of agility and getting this up I feel like it makes a ton of sense to include this in the React subfolder 😄 
- I've tried to rephrase the `info` section immediately after the `msw-trpc` installation to highlight that it is a new and unsupported package that folks should review and I've made the alternative to using this library much more explicit, pointing folks who want to avoid the package towards the [RPC spec documentation](https://trpc.io/docs/rpc)
- I super appreciate that tons of unit tests are not in vogue. My intention was to write a good doc to explain how to mock and I hope I've done a good job of emphasizing that this is primarily about being able to **isolate** frontend components, not **test** them. I've intentionally placed testing implementation for this reason. Personally, I continue to like to have some unit tests, so if someone would like to write some philosophical points in the introduction about alternatives I will not only support that but also read that with interest 😄 Thanks for the article link!

@Nick-Lucas I'm really sorry, I updated `sidebars.js` in my last commit but the `www` build failed and I can't see its errors on Vercel. Could you let me know if I'm overlooking anything? First time using Docusaurus to add a page!

File: www/docs/reactjs/mocking.md
Comment (juliusmarminge) [inline www/docs/reactjs/mocking.md:5]:
```suggestion
slug: /reactjs/mocking
```

File: www/docs/reactjs/mocking.md
Comment (c-ehrlich) [inline www/docs/reactjs/mocking.md:13]:
```suggestion
- Present components in frontend workshops like Storybook exactly as they would appear and behave in your main app
```

File: www/docs/reactjs/mocking.md
Comment (c-ehrlich) [inline www/docs/reactjs/mocking.md:58]: `createClient` can't be called without providing a transformer

Comment (c-ehrlich) [inline www/docs/reactjs/mocking.md:58]: `createClient` also needs to be passed a transformer
(note: this is also missing in trpc/react-query docs pages. maybe a recent type change?)

Comment (Nick-Lucas) [reply www/docs/reactjs/mocking.md:58]: Since transformer is optional, isn't it best to keep the docs boiled down to their simplest example? I think that's been the way so far in the docs and reduces visual noise for new learners

**Comment by Nick-Lucas** [comment]
> @Nick-Lucas I'm really sorry, I updated `sidebars.js` in my last commit but the `www` build failed and I can't see its errors on Vercel. Could you let me know if I'm overlooking anything? First time using Docusaurus to add a page!

You can probably get any failures like this by running `pnpm build-www` - it takes a few minutes but gives a pretty comprehensive list of problems at the end. Vercel build failure in my experience is always from this step

File: www/docs/reactjs/mocking.md
Comment (c-ehrlich) [reply www/docs/reactjs/mocking.md:58]: as far as i can tell its not possible to not pass a transformer here 
image

File: www/docs/client/mocking.md
Comment (c-ehrlich) [reply www/docs/client/mocking.md]: It seems to me like it will be fairly easy to trigger this issue, for example in a component that fires two different queries when it mounts. If at all possible we should try to get this working.

**Comment by c-ehrlich** [comment]
I've been trying for the past hour or so to get a basic `msw-trpc` example repo going. I started from `create-t3-app@latest` and followed the instructions from the PR but keep running into type errors, and even when I ignore them the test fails. The type errors seem to lead to a type error in the package.json of `msw-trpc`.

image

Here's my repo in broken state (see `src/__test__`): https://github.com/c-ehrlich/mock

If the issue is something specific to create-t3-app then that should be fixed on our side, otherwise I'm curious what I'm doing wrong.

**Comment by jharlowuk** [comment]
Hey @c-ehrlich! Thanks for taking the time to create that repo, it will be so valuable to have a simple demo repo to go along with the documentation.

Screen Shot 2023-03-19 at 6 40 06 PM

I was able to get the test passing based on a successful mock in your repo. I've made a [pull request](https://github.com/c-ehrlich/mock/pull/1) on it with my changes. I want to highlight a couple of things:
1. I think the only change I had to make in the `setup.tsx` was to match the client transformers with the `trpcMsw` transformers. This was a really valuable exercise and I will update this article this reflect this.
2. I added `jest-dom` and `node-fetch` I will be 100% honest and say that I don't know why my T3 app handles `global.fetch` fine but I have to import it here. Would really appreciate you explaining this if you know why that needs to be done manually.
3. **I am 100% unsatisfied with the typing coming from `ctx` in `trpcMsw`, and my solution feels very flakey** ([pulled from this existing issue](https://github.com/maloguertin/msw-trpc/issues/2)). You'll see a lot of lining errors still and I will try and spend some time in the next couple of weeks seeing if I can figure out a fix in the `msw-trpc`. Would appreciate another set of eyes on my solution though in case I missed anything 👀

**Comment by c-ehrlich** [comment]
> **Update: Don't spend any time on this, I figured it out but will write it up and update this tomorrow**
> 
> Hey @c-ehrlich! Thanks for taking the time to create that repo, it will be so valuable to have a simple demo repo to go along with the documentation.
> 
> Screen Shot 2023-03-19 at 6 40 06 PM
> 
> I was able to get the test passing based on a successful mock in your repo. I've made a [pull request](https://github.com/c-ehrlich/mock/pull/1) on it with my changes. I want to highlight a couple of things:
> 
>     1. I think the only change I had to make in the `setup.tsx` was to match the client transformers with the `trpcMsw` transformers. This was a really valuable exercise and I will update this article this reflect this.
> 
>     2. I added `jest-dom` and `node-fetch` I will be 100% honest and say that I don't know why my T3 app handles `global.fetch` fine but I have to import it here. Would really appreciate you explaining this if you know why that needs to be done manually.
> 
>     3. **I am 100% unsatisfied with the typing coming from `ctx` in `trpcMsw`, and my solution feels very flakey** ([pulled from this existing issue](https://github.com/maloguertin/msw-trpc/issues/2)). You'll see a lot of lining errors still and I will try and spend some time in the next couple of weeks seeing if I can figure out a fix in the `msw-trpc`. Would appreciate another set of eyes on my solution though in case I missed anything 👀

Making progress!

I managed to get `httpLink` working which feels much better than having to use a problematic `httpBatchLink` in tests. Also made a couple other minor changes.

remaining issues as far as I can tell:
- having to patch fetch - no idea why this is necessary either. but not doing it breaks the tests. could you link the repo of your T3 app where it doesn't need to be patched?
- `req.getInput()` should ideally just know that we're using superjson so we don't have to manually grab `.json` from it. this is how normal tRPC works so I'm sure it's possible
- `ctx.data` should be typed properly
- once this minimal sample is working i'd like to see a slightly bigger project, or maybe even a different setup, for example a Fastify/React monorepo or something. just to make sure that we're shipping a docs page that will work for most people from the start.

**Comment by jharlowuk** [comment]
Hey @c-ehrlich 

Lovely seeing your commits on mock today. I think you did pretty much exactly the same thing as me. mostly it was switching the batch link out for an `httpLink` 😄 

I've played around with `msw-trpc` a little more today and I respect what they've done but I feel like there's got to be a better/simpler way to type it. I'm going to do some playing this week and see if I can come up with something. I got `getInput` to type properly this evening but not in a way I like.

**Comment by jharlowuk** [comment]
Hey folks, apologies work has gotten very busy this week. I don't know when I can commit to look into `msw-trpc` more deeply, but I wonder if before making a more significant patch to that library to fix the input typing (which I am confident can be done, just trying to figure out the best approach), we can find a version of this page that we feel gives value in the interim?

I've made a couple of changes to the page ([see this commit](https://github.com/trpc/trpc/pull/4017/commits/1807ed14b40a60e14c7d664a758c73568ed26d5c)) to reflect your feedback regarding transformers, what works in my version of @c-ehrlich's `mock` repo, and highlighting a way to debug the captured `msw` request.

**Comment by maloguertin** [comment]
> Hey folks, apologies work has gotten very busy this week. I don't know when I can commit to look into `msw-trpc` more deeply, but I wonder if before making a more significant patch to that library to fix the input typing (which I am confident can be done, just trying to figure out the best approach), we can find a version of this page that we feel gives value in the interim?
> 
> I've made a couple of changes to the page ([see this commit](https://github.com/trpc/trpc/pull/4017/commits/1807ed14b40a60e14c7d664a758c73568ed26d5c)) to reflect your feedback regarding transformers, what works in my version of @c-ehrlich's `mock` repo, and highlighting a way to debug the captured `msw` request.

Hi guys! I am the author of the msw-trpc, sorry I've been away for some time and missed this PR. I have a PR on my package to improve the input typings with transformers: 

https://github.com/maloguertin/msw-trpc/pull/10

I haven't had time to dig deeply into all the comments here but I am more than willing to help.

**Comment by maloguertin** [comment]
> 3\. ** am 100% unsatisfied with the typing coming from `ctx` in `trpcMsw`, and my solution feels very flakey** ([pulled from this existing issue](https://github.com/maloguertin/msw-trpc/issues/2)). You'll see a lot of lining errors still and I will try and spend some time in the next couple of weeks seeing if I can figure out a fix in the `msw-trpc`. Would appreciate another set of eyes on my solution though in case I missed anything

Could you give me a bit more details on this? I've been trying to get a repo to reproduce the issues but to no avail.

**Comment by jharlowuk** [comment]
Hey @maloguertin! Thanks for creating such an awesome tool, I'm using `trpc-msw` every day. I looked at your PR and that's a way nicer solution to the `.getInput()` and `ctx.data()` typing than what I tried/alluded to, and based on a quick pull and integration with my current repo, it looks like it's fixed the issue! Looking forward to having `getInput()` on my mutations when you merge this 😄 

If you still want to have a look at the problem it should have linting errors in @c-ehrlich's [repo](https://github.com/c-ehrlich/mock).

**Comment by maloguertin** [comment]
> I just released v1.3.0 which properly handles the transformer and should fix the context bug. (hopefully 🤞)

**Comment by c-ehrlich** [comment]
> I just release v1.3.0 which properly handles the transformer and should fix the context bug. (hopefully 🤞)

nice, and thanks for all your work on this package so far! excited to get this into the docs soon :)

i still can't seem to satisfy all the types. could you either check out this https://github.com/c-ehrlich/mock and make a pr, or provide a minimal sample repo of your own?

**Comment by maloguertin** [comment]
> > I just release v1.3.0 which properly handles the transformer and should fix the context bug. (hopefully 🤞)
> 
> nice, and thanks for all your work on this package so far! excited to get this into the docs soon :)
> 
> i still can't seem to satisfy all the types. could you either check out this https://github.com/c-ehrlich/mock and make a pr, or provide a minimal sample repo of your own?

I'll look at your repo!

**Comment by jharlowuk** [comment]
Hey @maloguertin I checked out your PR and it seems great to me.

@c-ehrlich: I've updated the doc page with the new .getInput() method, but could I get a review? If you can't give an approval yet, I'd love a clear sense of what needs to be done to get this there! <3

**Review by c-ehrlich — CHANGES_REQUESTED** [review]
> Hey @maloguertin I checked out your PR and it seems great to me.
> 
> @c-ehrlich: I've updated the doc page with the new .getInput() method, but could I get a review? If you can't give an approval yet, I'd love a clear sense of what needs to be done to get this there! <3

What's still needed is a working demo - in the long run there should probably be an example project within the trpc repo that shows all the different types of tests, but for now a repo anywhere that we can link to with one mocked api call is good enough to merge IMO. Docs themselves are good.

I tried again to get the whole thing set up based on the docs. There's huge progress: the new msw-trpc stuff is great, and there's not a single type error. But my test is failing, seems like the mock isn't intercepting the request. Probably some dumb mistake on my side? https://github.com/c-ehrlich/mock2

Also a nit but maybe its worth making all the imports explicit in the docs? (but happy to approve without)

File: www/docs/reactjs/mocking.md
Comment (juliusmarminge) [inline www/docs/reactjs/mocking.md:22]: _app.tsx

**Comment by maloguertin** [comment]
> > Hey @maloguertin I checked out your PR and it seems great to me.
> > @c-ehrlich: I've updated the doc page with the new .getInput() method, but could I get a review? If you can't give an approval yet, I'd love a clear sense of what needs to be done to get this there! <3
> 
> What's still needed is a working demo - in the long run there should probably be an example project within the trpc repo that shows all the different types of tests, but for now a repo anywhere that we can link to with one mocked api call is good enough to merge IMO. Docs themselves are good.
> 
> I tried again to get the whole thing set up based on the docs. There's huge progress: the new msw-trpc stuff is great, and there's not a single type error. But my test is failing, seems like the mock isn't intercepting the request. Probably some dumb mistake on my side? https://github.com/c-ehrlich/mock2
> 
> Also a nit but maybe its worth making all the imports explicit in the docs? (but happy to approve without)

I cloned your repo, I had to install `whatwg-fetch` and import it in setup.tsx and the tests we're passing.

**Comment by c-ehrlich** [comment]
> I cloned your repo, I had to install `whatwg-fetch` and import it in setup.tsx and the tests we're passing.

Got it working! Was getting some type errors with `whatwg-fetch` so I used `cross-fetch` instead. But no preferences as to what the docs recommend. I renamed the repo: https://github.com/c-ehrlich/msw-trpc-example

Will read the PR over tomorrow :)

File: www/docs/reactjs/mocking.md
Comment (c-ehrlich) [inline www/docs/reactjs/mocking.md:203]: Maybe something like

```suggestion
First we render our `<BestCat />` component with the necessary context using the `renderWithProviders` function that we created earlier. Then we assert that the component will start out in loading state, as it needs to asynchronously wait for the data to load. Once we have waited for the data to load, we assert on the anticipated response.
```
?

File: www/docs/reactjs/mocking.md
Comment (c-ehrlich) [inline www/docs/reactjs/mocking.md:205]:
```suggestion
The component is now successfully making calls from its tRPC client when being tested. But we don't want it to make those calls to a real server. This is where our mocked `server` comes in. Inside of `setupServer`, we can define the requests that our `Cat` component uses so we can intercept them and return a mocked response. Because `trpcMsw` uses the shape of our `AppRouter`, we can use our existing typing to quickly define the request handler we need.
```

File: www/docs/reactjs/mocking.md
Comment (c-ehrlich) [inline www/docs/reactjs/mocking.md:216]: Add a link to a working repo here? The one I posted yesterday would work, but even better would be one with dependabot, a ci pipeline, etc so it stays up to date. Could even add this to `/examples`

**Review by c-ehrlich — COMMENTED** [review]
All of these are suggestions, feel free to merge without addressing. The only thing I'd like to have for sure is a link to a fully working sample.

**Comment by KATT** [comment]
cc @c-ehrlich do you wanna take over this one? :)

**Comment by KATT** [comment]
If you don't use it already, [GitHub's CLI](https://cli.github.com/) is great, allows you to do `gh pr checkout 4017`

**Comment by juliusmarminge** [comment]
Superseded by #4451

**Pull request closed without merging**
